stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+        """
+        if not self.resource:
+            if not self.exists():
+                raise ValueError('Project ({}) cannot be created with makedirs_p'.format(self))
+            return
+        proj_handler = dxpy.DXProject(self.canonical_project)
+        proj_handler.new_folder(self.resource, parents=True)
 
     def isdir(self):
-        if not self.resource or self._is_folder():
-            return self.exists()
-        return False
+        """Determine if path is an existing directory
+
+        Returns (bool) : True if path is an existing folder path or project
+        """
+        if not self.resource and self.exists():  # path could be a project
+            return True
+        # or path could be a folder
+        try:
+            self.listdir()
+            return True
+        except NotFoundError:
+            return False
 
     def isfile(self):
+        """Determine if path is an existing file
+
+        Returns (bool) : True if path points to an existing file
+        """
+        if not self.resource or self.endswith('/'):
+            return False
         try:
-            return self.resource and not self._is_folder() and self.exists()
+            self.stat()
+            return True
         except NotFoundError:
             return False
 
-    def getsize(self):
+    def rename(self, new_name):
+        """Rename a data object on the DX platform
+
+        Args:
+            new_name (str): New name of the object
+
+        Raises:
+            ValueError : When trying to rename a project
+        """
         if not self.resource:
-            return self.stat()['dataUsage']*1e9
-        elif self._is_folder():
-            return 0
+            raise ValueError('Cannot rename project ({})'.format(self))
+        if new_name == self.name:
+            return
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        file_handler.rename(new_name)
+
+    def _clone(self, dest):
+        """Clones the data object into the destination path.
+        The original file is retained.
+
+        Args:
+            dest (path) : The destination file/folder path in a different project
+
+        Raises:
+            ValueError : If attempting to clone a project
+            DNAnexusError : If cloning within same project
+        """
+        if not self.resource:
+            raise ValueError('Cannot clone project ({})'.format(self))
+        if dest.canonical_project == self.canonical_project:
+            raise DNAnexusError('Cannot clone within same project')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        dest_exists = dest.isdir()
+        if dest_exists:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise DuplicateError('Dest path already exists. Do not duplicate!')
```

Comment:
why is this error raised before the copy?

Additionally, prob clearer to say `Destination path %s already exists, will not cause duplicate file objects to exist. Remove the original first`.

Reply:
Suppose you have an existing file `proj1:/folder1/file.txt` and you want to move `file.txt` from some source into `proj1:/folder`. Since folder dir already exists, we try to move it into `proj1:/folder/file.txt`, but that exists too! So we raise a DuplicateError before copy.

Similarly, if say a file `proj1:/file2.txt` exists and we are moving a source file `file.txt` into `proj1:/file2.txt`, we see that file2.txt is not an existing dir, so we try to copy the file into `proj1:/file2.txt` but this already exists! We raise an error in this scenario too.

Reply:
Changed the error message to reflect yours though.